src/main/scala/com/spotify/ratatool/samplers/BigSampler.scala

```diff
@@ -0,0 +1,346 @@
+    fields.foldLeft(hasher) { case (h, f) =>
+      val fieldParts = f.split('.')
+      val fieldOpt = tblSchema.find(_.getName == fieldParts.head)
+      if (fieldOpt.isEmpty) {
+        log.warn(s"Can't find field ${fieldParts.head} in table schema")
+        hasher
+      } else {
+        val v = r.get(fieldParts.head)
+        val field = fieldOpt.get
+        if (field.getMode == "REPEATED") {
+          //TODO: support repeated fields
+          throw new UnsupportedOperationException(
+            "Repeated field is not supported as sampling key!")
+        }
+        field.getType match {
+          case "BOOLEAN" => h.putBoolean(v.asInstanceOf[Boolean])
+          case "INTEGER" => h.putLong(v.asInstanceOf[Long])
+          case "FLOAT" => h.putFloat(v.asInstanceOf[Float])
+          case "STRING" => h.putString(v.asInstanceOf[String], utf8Charset)
+          case "BYTES" => h.putBytes(v.asInstanceOf[ByteString].toByteArray)
+          case "TIMESTAMP" =>
+            throw new UnsupportedOperationException("Timestamp is not supported as sampling key!")
+          case "DATE" =>
+            throw new UnsupportedOperationException("Date is not supported as sampling key!")
+          case "TIME" =>
+            throw new UnsupportedOperationException("Time is not supported as sampling key!")
+          case "DATETIME" =>
+            throw new UnsupportedOperationException("Datetime is not supported as sampling key!")
+          case "RECORD" =>
+            hashTableRow(
+              v.asInstanceOf[TableRow],
+              field.getFields.asScala.toList,
+              fieldParts.tail,
+              h)
+          case t => throw new IllegalArgumentException(s"Type: $t not supported")
+        }
+      }
+    }
+  }
+  // scalastyle:on cyclomatic.complexity
+
+  private def sampleBigQueryTable(sc: ScioContext,
+                                  inputTbl: TableReference,
+                                  outputTbl: TableReference,
+                                  fields: List[String],
+                                  samplePct: Float,
+                                  seed: Option[Int]): Future[Tap[TableRow]] = {
+    val patchedBigQueryService = new PatchedBigQueryService(sc.optionsAs[BigQueryOptions])
+
+    if (patchedBigQueryService.getTable(outputTbl) != null) {
+      log.info(s"Reuse previous sample at $outputTbl")
+      Taps().bigQueryTable(outputTbl)
+    } else {
+      log.info(s"Will sample from BigQuery table: ${inputTbl}, output will be ${outputTbl}")
+
+      val schema = patchedBigQueryService.getTable(inputTbl).getSchema
+
+      // defeat closure TableSchema and TableFieldSchema are not serializable
+      val schemaStr = JsonSerDe.toJsonString(schema)
+      @transient lazy val schemaFields =
+        JsonSerDe.fromJsonString(schemaStr, classOf[TableSchema]).getFields.asScala.toList
+      val (s, n) = samplePctToInt(samplePct, 100)
+      val r = sc.bigQueryTable(inputTbl)
```

Comment:
Was thinking of this, but decided against it for now, as it was simpler to keep it in scala for now. Might revisit later if this is a bottleneck. Does it make sense?